GitCommands/Git/GitRevisionTester.cs

```diff
@@ -1,15 +1,106 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.IO.Abstractions;
 using System.Linq;
 
 namespace GitCommands.Git
 {
     public interface IGitRevisionTester
     {
+        /// <summary>
+        /// Finds if all of the first selected are parents to the selected revision
+        /// </summary>
+        /// <param name="firstSelected">The first selected revisions (A)</param>
+        /// <param name="selectedRevision">The currently (last) selected revision (B)</param>
+        /// <returns>
+        /// True if one of the first selected is parent
+        /// </returns>
+        bool AllFirstAreParentsToSelected(IEnumerable<GitRevision> firstSelected, GitRevision selectedRevision);
+
+        /// <summary>
+        /// Finds if any of the git items exists as a file.
+        /// </summary>
+        /// <param name="selectedItemsWithParent">List of items to resolve and check</param>
+        /// <returns>
+        /// True if at least one file exists.
+        /// </returns>
+        bool AnyLocalFileExists(IEnumerable<GitItemStatus> selectedItemsWithParent);
+
         bool Matches(GitRevision revision, string criteria);
     }
 
     public class GitRevisionTester : IGitRevisionTester
     {
+        private readonly IFullPathResolver _fullPathResolver;
+        private readonly IFileSystem _fileSystem;
+
+        public GitRevisionTester(IFullPathResolver fullPathResolver, IFileSystem fileSystem = null)
+        {
+            _fullPathResolver = fullPathResolver;
+            _fileSystem = fileSystem ?? new FileSystem();
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Finds if all of the first selected are parents to the selected revision
+        /// </summary>
+        /// <param name="firstSelected">The first selected revisions (A)</param>
+        /// <param name="selectedRevision">The currently (last) selected revision (B)</param>
+        /// <returns>
+        /// True if one of the first selected is parent
+        /// </returns>
+        public bool AllFirstAreParentsToSelected(IEnumerable<GitRevision> firstSelected, GitRevision selectedRevision)
+        {
+            if (selectedRevision?.ParentGuids == null)
+            {
+                return false;
+            }
+
+            foreach (var item in firstSelected.Select(r => r.Guid))
+            {
+                if (!selectedRevision.ParentGuids.Contains(item, StringComparer.Ordinal))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
```

Comment:
We need to rename the method to `AnyFirstAreParentsToSelected` to match the implementation and the doc "_True if **one of** the first selected is parent_".

Reply:
:bulb:  Thinking of it, may be we can change the signature to something like the following? 

```
+        /// <summary>
+        /// Indicates whether any of the provided <param ref="selectedRevision" />'s parents are selected.
+        /// </summary>
+        public bool AnyRevisionParentsSelected(GitRevision selectedRevision, IEnumerable<GitRevision> revisionsSelectedFirst)
```

I think API becomes slightly more readable:

```
bool firstIsParent = _revisionDiffController.AnyRevisionParentsSelected(DiffFiles.Revision, DiffFiles.SelectedItemParents);
 ````



A rough idea, open to discussion...

Reply:
The method checks if "True if all of the first selected are parents" so this would be changed behavior

Reply:
My bad, misread the logic.

Reply:
I will be happy to take a discussion why I have done like I have done and to change it to something better. How should this work? Your comment is a good indication of that. I could give some more background tomorrow. The alternative is to take this in follow-up discussions.

Reply:
That's be good. I'm finding API somewhat confusing.

Reply:
This PR (including the related PRs) primarily improves the diff and presentation of versions. In this process, the menu logic has been updated, some alternatives were not relevant and other were missing. The logic was also changed so testing could be done (testing is for the display, not for the actions so I do not find the tests useful still but they suck less than they would in my first iterations).

For the alternatives, some kind of rules were used:
 * With one item selected, the menu items must be relevant
 * With multiple items selected, menu items are relevant if at least one item is relevant. For instance if submodules are available if at least one submodule is selected. (There are exceptions to this rule)

In general, the behavior should be the same if the parent-child is selected as if just the "child". It was not like this previously, selecting a parent-child was more limiting. 
In some situations selecting the parent, then the selected will provide more information, like the actual parents to parents, that enables those menu items (in diff menu).
AllFirstAreParentsToSelected() is a helper for these situations, this is similar to just selecting one revision.

There are some special handling for merge commits, to show the combined diff like in e51741dcbf694d937db2c0fbeabf107827562fa7. This selects all parents and adds the artificial DiffFiles.CombinedDiff to compare to the Selected revision. The combined diff is shown if only one of the parents to e517 is shown but would be confusing if a non parent was selected. 
(So combined diff is shown if at least one parent shown, but no other revision than parents).

For staged/unstaged, the availability is available is based on the selected revision but the parent must be staged/Head to be sure a file can be staged/unstaged. This commands breaks the rule to "show if at least one is possible" as it is not possible to determine the availability in a good way for the command.

FirstIsParent is also used to confirm that a parent to parent exists for difftool exists and to prune when parent-to-selected(B) is is the same as first(A).

Pushing a few clarifications related to this.

Not sure if it got any clearer...